sdk/python/feast/infra/online_stores/tikv.py

```diff
@@ -0,0 +1,174 @@
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
+
+import pytz
+from pydantic import StrictStr
+from pydantic.schema import Literal
+
+from feast import Entity, FeatureTable
+from feast.feature_view import FeatureView
+from feast.infra.key_encoding_utils import serialize_entity_key
+from feast.infra.online_stores.online_store import OnlineStore
+from feast.protos.feast.types.EntityKey_pb2 import EntityKey as EntityKeyProto
+from feast.protos.feast.types.Value_pb2 import Value as ValueProto
+from feast.repo_config import FeastConfigBaseModel, RepoConfig
+
+# TODO: release https://github.com/tikv/client-py
+from tikv_client import RawClient
+
+class TiKVOnlineStoreConfig(FeastConfigBaseModel):
+    """ Online store config for local (TiKV-based) store """
+
+    type: Literal["tikv", "feast.infra.online_stores.tikv.TiKVOnlineStore"] = "tikv"
+    """ Online store type selector"""
+
+    mode: StrictStr = "rawkv"
+    """KV mode for TiKV
+     format: rawkv or txnkv """
+
+    pd_addresses: StrictStr = "127.0.0.1:2379"
+    """PD Addresses for the TiKV Cluster
+     format: host1:port,host2:port,host3:port eg. pd1:2379,pd2:2379,pd3:2379 """
+
+
+class TiKVOnlineStore(OnlineStore):
+    """
+    OnlineStore is an object used for all interaction between Feast and the service used for offline storage of
+    features.
+    """
+
+    _rawkv_client: Optional[RawClient] = None
+
+    @staticmethod
+    def _is_rawkv(online_store_config: TiKVOnlineStoreConfig) -> bool:
+        return online_store_config.mode.lower() == "rawkv"
+
+    @staticmethod
+    def _encode_tikv_key(
+        project: str, entity_key: EntityKeyProto, feature_view: str, feature_name: str
+    ):
+        # TODO: optimize encode key
+        entity_key_bin = serialize_entity_key(entity_key)
+        tikv_key = (
+            f"{project}:".encode("utf8")
+            + entity_key_bin
+            + f":{feature_view}:{feature_name}".encode("utf8")
+        )
+        return tikv_key
+
+    def _get_rawkv_client(self, online_store_config: TiKVOnlineStoreConfig):
+        if not self._rawkv_client:
+            _rawkv_client = RawClient.connect(online_store_config.pd_addresses)
+        return _rawkv_client
+
+    def update(
+        self,
+        config: RepoConfig,
+        tables_to_delete: Sequence[Union[FeatureTable, FeatureView]],
+        tables_to_keep: Sequence[Union[FeatureTable, FeatureView]],
+        entities_to_delete: Sequence[Entity],
+        entities_to_keep: Sequence[Entity],
+        partial: bool,
+    ):
+        """
+        There's currently no setup done for TiKV.
+        """
+        pass
+
+    def teardown(
+        self,
+        config: RepoConfig,
+        tables: Sequence[Union[FeatureTable, FeatureView]],
+        entities: Sequence[Entity],
+    ):
+        """
+        There's currently no teardown done for TiKV.
+        """
+        pass
+
+    def online_write_batch(
```

Comment:
maybe it's better to use bulk load API if the data to be written is too huge.

Reply:
It's better to use bulk load API in https://github.com/feast-dev/feast-spark.